numba_dppy/utils/array_utils.py

```diff
@@ -0,0 +1,205 @@
+
+    Args:
+        usm_backed: Object conformant to __sycl_usm_array_interface__.
+        obj (numpy.ndarray): Numpy ndarray, the data will be copied into.
+
+    Raises:
+        TypeError: If any argument is not of permitted type.
+        ValueError: If size of data does not match.
+    """
+    usm_mem = is_usm_backed(usm_backed)
+    if usm_mem is None:
+        raise TypeError("Source is not USM backed.")
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_from_host(obj_memview)
+
+
+def copy_from_usm_backed(usm_backed, obj):
+    """
+    Copy from USM backed data to supported objects.
+
+    Args:
+        usm_backed: Object conformant to __sycl_usm_array_interface__.
+        obj (numpy.ndarray): Numpy ndarray, the data will be copied into.
+
+
+    Raises:
+        TypeError: If any argument is not of permitted type.
+        ValueError: If size of data does not match.
+    """
+    usm_mem = is_usm_backed(usm_backed)
+    if usm_mem is None:
+        raise TypeError("Source is not USM backed.")
+
+    if not isinstance(obj, np.ndarray):
+        raise TypeError(
+            "Obj is not USM backed and is not of type "
+            "numpy.ndarray. Obj type: %s" % (type(obj))
+        )
+
+    if obj.dtype not in [np.dtype(typ) for typ in supported_numpy_dtype]:
+        raise ValueError(
+            "dtype is not supprted. Supported dtypes "
+            "are: %s" % (supported_numpy_dtype)
+        )
+
+    size = np.prod(obj.shape)
+    if usm_mem.size != (obj.dtype.itemsize * size):
+        raise ValueError(
+            "Size (Bytes) of data does not match. USM backed "
+            "memory size %d, supported object size: %d"
+            % (usm_mem.size, (obj.dtype.itemsize * size))
+        )
+
+    obj_memview = memoryview(obj)
+    obj_memview = obj_memview.cast("B")
+    usm_mem.copy_to_host(obj_memview)
+
+
+def as_usm_backed(obj, queue=None, usm_type="shared", copy=True):
+    """
+    Determine and return a SYCL device accesible object.
+
+    We try to determine if the provided object conforms to
+    __sycl_usm_array_interface__. If not, we create a USM memory of `usm_type`
+    and try to copy the data obj holds. If obj is not already backed by USM,
```

Comment:
Add backticks around `obj`. I could not parse the last line, seems to be a broken sentence.